server/src/main/java/org/candlepin/policy/js/autobind/AutobindRules.java

```diff
@@ -68,29 +67,38 @@ public AutobindRules(JsRunner jsRules, ProductCurator productCurator, RulesObjec
         List<Pool> pools, ComplianceStatus compliance, String serviceLevelOverride,
         Set<String> exemptLevels, boolean considerDerived) {
 
+        List<PoolQuantity> bestPools = new ArrayList<PoolQuantity>();
         int poolsBeforeContentFilter = pools.size();
         pools = filterPoolsForV1Certificates(consumer, pools);
         log.debug("pools.size() before V1 certificate filter: {}, after: {}",
             poolsBeforeContentFilter, pools.size());
 
+        List<String> productIdList;
+        if (productIds == null) {
+            productIdList = new ArrayList<String>();
+        }
+        else {
+            productIdList = Arrays.asList(productIds);
+        }
+        List<String> consumerInstalledProducts = new ArrayList<String>();
+        for (ConsumerInstalledProduct cip : consumer.getInstalledProducts()) {
+            consumerInstalledProducts.add(cip.getProductId());
+        }
+
         if (pools.size() == 0) {
             if (compliance.getReasons().size() == 0) {
                 log.info("Consumer is compliant and does not require more entitlements.");
             }
             else {
-                Set<String> fullList = new HashSet<String>();
-                fullList.addAll(Arrays.asList(productIds));
-                for (ConsumerInstalledProduct cip : consumer.getInstalledProducts()) {
-                    fullList.add(cip.getProductId());
-                }
-                log.info("No pools available to complete compliance for the set of proudcts: " + fullList);
+                log.info("No pools available to complete compliance for the set of products: {}" +
```

Comment:
```
if (log.isInfoEnabled()) {
  List<String> cips = new LinkedList<String>();
  for (ConsumerInstalledProduct cip : consumer.getInstalledProducts()) {
    cips.add(cip.getProductId());
  }

  log.info("No pools available...", productIdList, cips);
}
```

Note the use of the linked list here rather than an array list. It's slightly more performant for appending and linear traversal. Also, if the log line won't be printed, none of the work of fetching the data (and possibly invoking a lazy lookup) is done.